core-bundle/src/Intl/Locales.php

```diff
@@ -0,0 +1,273 @@
+     */
+    public function getLocaleIds(): array
+    {
+        // If the legacy hook is used, it might add or remove locales
+        if (!empty($GLOBALS['TL_HOOKS']['getLanguages'])) {
+            $locales = array_keys($this->getLocales('en'));
+            sort($locales);
+
+            return $locales;
+        }
+
+        return $this->getLocaleIdsWithoutHook();
+    }
+
+    /**
+     * @return array<string> Languages (without regions) as ICU locale IDs
+     */
+    public function getLanguageLocaleIds(): array
+    {
+        // If the legacy hook is used, it might add or remove locales
+        if (!empty($GLOBALS['TL_HOOKS']['getLanguages'])) {
+            $locales = array_keys($this->getLanguages('en'));
+            sort($locales);
+
+            return $locales;
+        }
+
+        $localeIds = array_map(
+            static function ($localeId) {
+                $locale = \Locale::parseLocale($localeId);
+
+                if (!isset($locale[\Locale::REGION_TAG])) {
+                    return $localeId;
+                }
+
+                unset($locale[\Locale::REGION_TAG]);
+
+                return \Locale::composeLocale($locale);
+            },
+            $this->getLocaleIdsWithoutHook()
+        );
+
+        $localeIds = array_values(array_unique(array_filter($localeIds)));
+
+        sort($localeIds);
+
+        return $localeIds;
+    }
+
+    /**
+     * @return array<string,string> Translated locales indexed by their ICU locale IDs
+     */
+    public function getDisplayNames(array $localeIds, string $displayLocale = null, bool $addNativeSuffix = false): array
+    {
+        $locales = $this->getDisplayNamesWithoutHook($localeIds, $displayLocale, $addNativeSuffix);
+
+        if (!empty($GLOBALS['TL_HOOKS']['getLanguages'])) {
+            return $this->applyLegacyHook($locales, true);
+        }
+
+        return $locales;
+    }
+
+    private function getDisplayNamesWithoutHook(array $localeIds, string $displayLocale = null, bool $addNativeSuffix = false): array
+    {
+        if (null === $displayLocale && null !== ($request = $this->requestStack->getCurrentRequest())) {
+            $displayLocale = $request->getLocale();
+        }
+
+        $locales = [];
+
+        foreach ($localeIds as $localeId) {
+            $langKey = 'LNG.'.$localeId;
+            $label = $this->translator->trans($langKey, [], 'contao_languages', $displayLocale);
+
+            if ($label === $langKey || !\is_string($label) || '' === $label) {
+                $label = \Locale::getDisplayName($localeId, $displayLocale ?? 'en');
+            }
+
+            if ($addNativeSuffix) {
+                $nativeLabel = \Locale::getDisplayName($localeId, $localeId);
+
+                if ($nativeLabel !== $label) {
+                    $label .= ' - '.$nativeLabel;
+                }
+            }
+
+            $locales[$localeId] = $label;
+        }
+
+        (new \Collator($displayLocale ?? 'en'))->asort($locales);
```

Comment:
Should we use the request locale as fallback?

Reply:
https://github.com/contao/contao/blob/8796a68c24a69e8f7c2e36c0ca205b684fe6d4ba/core-bundle/src/Intl/Locales.php#L178